filters/RelaxationDartThrowing.cpp

```diff
@@ -0,0 +1,157 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "RelaxationDartThrowing.hpp"
+
+#include <pdal/KDIndex.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+#include <pdal/util/Utils.hpp>
+
+#include <chrono>
+#include <numeric>
+#include <random>
+#include <string>
+#include <vector>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{
+    "filters.relaxationdartthrowing", "Subsampling filter",
+    "http://pdal.io/stages/filters.relaxationdartthrowing.html"};
+
+CREATE_STATIC_STAGE(RelaxationDartThrowing, s_info)
+
+std::string RelaxationDartThrowing::getName() const
+{
+    return s_info.name;
+}
+
+void RelaxationDartThrowing::addArgs(ProgramArgs& args)
+{
+    args.add("decay", "Decay rate", m_decay, 0.9);
+    args.add("radius", "Minimum radius (initial)", m_startRadius, 1.0);
+    args.add("count", "Target number of points after sampling", m_maxSize,
+             (point_count_t)1000);
+    args.add("shuffle", "Shuffle points prior to sampling?", m_shuffle, true);
+    m_seedArg = &args.add("seed", "Random number generator seed", m_seed);
+}
+
+PointViewSet RelaxationDartThrowing::run(PointViewPtr inView)
+{
+    point_count_t np = inView->size();
+
+    // Return empty PointViewSet if the input PointView has no points.
+    // Otherwise, make a new output PointView.
+    PointViewSet viewSet;
+    if (!np)
+        return viewSet;
+    PointViewPtr outView = inView->makeNew();
+
+    // Build the 3D KD-tree.
+    KD3Index& index = inView->build3dIndex();
+
+    PointIdList finalIds;
+
+    double radius(m_startRadius);
+
+    PointIdList shuffledIds(np);
+    std::iota(shuffledIds.begin(), shuffledIds.end(), 0);
+    if (m_shuffle)
+    {
+        if (!m_seedArg->set())
+            m_seed =
+                std::chrono::system_clock::now().time_since_epoch().count();
+        std::shuffle(shuffledIds.begin(), shuffledIds.end(),
+                     std::mt19937(m_seed));
+    }
+
+    while (finalIds.size() < m_maxSize)
+    {
+        // All points are marked as kept (1) by default. As they are masked by
+        // neighbors within the user-specified radius, their value is changed to
+        // 0.
+        std::vector<int> keep(np, 1);
+
+        // Start by masking all points within radius of finalIds.
+        for (PointId i : finalIds)
+        {
+            PointIdList ids = index.radius(i, radius);
+            for (PointId const& id : ids)
+                keep[id] = 0;
+        }
+
+        // We are able to subsample in a single pass over the shuffled indices.
+        for (PointId const& i : shuffledIds)
```

Comment:
Could you maybe get into an infinite loop here? If you don't hit the number of points in `finalIds` that you want, but you have eliminated all the nearby points by setting keep to 0?

Reply:
If you do not hit the target number of points, the mask is reset and the radius decreased. I think the only situations that could get you in trouble are 1) requesting more output points than are in the input, and 2) if there are duplicate points such that shrinking the radius never discriminates them. Both of these feel like edge cases, but we could add some checks on them. Otherwise, I think the worst case is that you (rightly) get back the input point cloud.

Reply:
I guess I misunderstood what was happening.

This seems very expensive, as a single pass with the `radius` will give you all the information you need: You know all the points within the radius of the target point. The distances are available from nanoflann, though we don't seem to expose that right now. I guess the tradeoff is memory, but if you're actually going to end up relaxing (which is the point of the filter), I would think you're going to overwhelm with CPU, as nanoflann searching is slow.

Also, maybe when you've iterated so many times, you just give up and throw everything in? I get that with most real-life data and good parameters, this probably won't happen, but it also seems like something that someone will file an issue for.

Reply:
You are absolutely right that we could just stash the distances and recall those in subsequent iterations. I'll take a look at this.

Reply:
I'm fine with merging this. If you want to enhance later, that's fine.

I just wonder about the name change for "filters.farthestpointsampling"

Reply:
I just pushed an update. Let me know what you think. It could probably be prettier or more succinct, but it seems to get the job done.

The name change for FPS should have been a correction in the docs only to match the filter name as implemented.